relay_test.go

```diff
@@ -1361,3 +1362,207 @@ func inspectFrames(rh *relaytest.StubRelayHost) chan relay.CallFrame {
+		// 		cf.Arg2Append([]byte("key"), []byte("value"))
+		// 	},
+		// 	modifyArg2: func(m map[string]string) {
+		// 		m["key"] = "value"
+		// 	},
+		// },
+		// {
+		// 	msg: "add large key/value",
+		// 	modifyFrame: func(cf relay.CallFrame, _ *relay.Conn) {
+		// 		cf.Arg2Append([]byte(largeKey), []byte(largeVal))
+		// 	},
+		// 	modifyArg2: func(m map[string]string) {
+		// 		m[largeKey] = largeVal
+		// 	},
+		// },
+	}
+
+	checksumTypes := []struct {
+		msg          string
+		checksumType ChecksumType
+	}{
+		{"none", ChecksumTypeNone},
+		{"crc32", ChecksumTypeCrc32},
+		{"farmhash", ChecksumTypeFarmhash},
+		{"crc32c", ChecksumTypeCrc32C},
+	}
+
+	payloadTests := []struct {
+		msg  string
+		arg2 map[string]string
+		arg3 []byte
+	}{
+		{
+			msg:  "no payload",
+			arg2: nil, // empty map
+			arg3: []byte{},
+		},
+		{
+			msg: "small payloads",
+			arg2: map[string]string{
+				"existingKey": "existingValue",
+			},
+			arg3: []byte("hello world"),
+		},
+		{
+			msg: "large payloads",
+			arg2: map[string]string{
+				"existingKey": "existingValue",
+			},
+			arg3: []byte(largePayload),
+		},
+	}
+
+	const (
+		format      = Thrift
+		noErrMethod = "EchoVerifyNoErr"
+		errMethod   = "EchoVerifyErr"
+	)
+
+	appErrTests := []struct {
+		msg        string
+		method     string
+		wantAppErr bool
+	}{
+		{
+			msg:        "no app error bit",
+			method:     noErrMethod,
+			wantAppErr: false,
+		},
+		{
+			msg:        "app error bit",
+			method:     errMethod,
+			wantAppErr: true,
+		},
+	}
+
+	for _, mt := range modifyTests {
+		for _, csTest := range checksumTypes {
+			t.Run(mt.msg+"/checksum="+csTest.msg, func(t *testing.T) {
+				// Create a relay that will modify the frame as per the test.
+				relayHost := relaytest.NewStubRelayHost()
+				relayHost.SetFrameFn(mt.modifyFrame)
+				opts := testutils.NewOpts().
+					SetRelayHost(relayHost).
+					SetRelayOnly()
+				testutils.WithTestServer(t, opts, func(t testing.TB, ts *testutils.TestServer) {
+					// Create a client that uses a specific checksumType.
+					clientOpts := testutils.NewOpts().SetChecksumType(csTest.checksumType)
+					client := ts.NewClient(clientOpts)
+					defer client.Close()
+
+					// Create a server echo verify endpoints (optionally returning an error).
+					for _, appErrTest := range appErrTests {
+						handler := echoVerifyHandler{
+							t:            t,
+							verifyFormat: format,
+							verifyCaller: client.ServiceName(),
+							verifyMethod: appErrTest.method,
+							appErr:       appErrTest.wantAppErr,
+						}
+						ts.Server().Register(raw.Wrap(handler), appErrTest.method)
+					}
+
+					ctx, cancel := NewContextBuilder(testutils.Timeout(time.Second)).
+						SetFormat(format).Build()
+					defer cancel()
+
+					// Make calls with different payloads and expected errors.
+					for _, aet := range appErrTests {
+						for _, tt := range payloadTests {
+							arg2Encoded := encodeThriftHeaders(t, tt.arg2)
```

Comment:
can we use subtests here as well so it's easier to run only specific test cases?

Reply:
Sure, added messages, it looks like:
```
=== RUN   TestRelayModifyArg2/no_change/checksum=crc32/with_relay/app_error_bit,no_payload
=== RUN   TestRelayModifyArg2/no_change/checksum=crc32/with_relay/app_error_bit,small_payloads
=== RUN   TestRelayModifyArg2/no_change/checksum=crc32/with_relay/app_error_bit,large_payloads
```